File: CR12/Lecture13-14/tensorproduct.m
function T = tensorproduct(X, k1, Y, k2)

	dimsize1 = size(X);
	pv1 = [1:ndims(X), k1];

	pv1(k1) = [];
	dimsize1(k1) = [];
	A = reshape(permute(X, pv1), numel(X)/size(X,k1), []);
	
	dimsize2 = size(Y);
	pv2 = [k2, 1:ndims(Y)];

	pv2(k2+1) = [];
	dimsize2(k2) = [];
	
	B = reshape(permute(Y, pv2), size(Y,k2), []);
	
	
	finaldim = [dimsize1 dimsize2];
	
	T = reshape(A*B, finaldim);

end

File: CR12-2024/Lecture14/tensorproduct.py
from unfolding import *

def tensorproduct(X, k1, Y, k2):
    dimsize1 = list(X.shape)
    dimsize2 = list(Y.shape)

    del dimsize1[k1]
    del dimsize2[k2]

    finaldim = dimsize1 + dimsize2

    X_k1 = unfolding(X, k1)
    Y_k2 = unfolding(Y, k2)

    XY = np.transpose(X_k1) @ Y_k2

    return np.reshape(XY, tuple(finaldim), order='F')

